### PR TITLE
Allow multiple files in upload window, give more control

### LINKED ISSUE #95: Uploading a single file should not hide the drop area
I think uploading a single file should keep the drop box enabled to allow you to also add a second/third etc. Right now if you want to upload multiple files but do not drag them all at the same time, you need to click import again and then select the 'add to existing' option

_Originally posted by @WardBrian in https://github.com/flatironinstitute/stan-playground/pull/80#discussion_r1655135279_

### PR DESCRIPTION
**Pull request by WardBrian**
Closes #95 

You can now drop multiple items in succession, and you are shown a list with the ability to remove items before clicking the create project buttons. 

image

### CODE REVIEW

File: gui/src/app/pages/HomePage/LoadProjectWindow.tsx [hunk 1]
```diff
@@ -89,58 +81,85 @@ const LoadProjectWindow: FunctionComponent<LoadProjectWindowProps> = ({
...
-      // replace the current project.
-      setShowReplaceProjectOptions(true);
-    } else {
-      // Otherwise, we just go ahead and import the files, replacing the
-      // entire project
-      importUploadedFiles({ replaceProject: true });
-    }
-  }, [filesUploaded, importUploadedFiles]);
+  const onUpload = useCallback(
+    (fs: { name: string; content: ArrayBuffer }[]) => {
```
Comment (jsoules) [inline gui/src/app/pages/HomePage/LoadProjectWindow.tsx:85]: Can use `File` type alias defined above

File: gui/src/app/pages/HomePage/LoadProjectWindow.tsx [hunk 2]
```diff
@@ -89,58 +81,85 @@ const LoadProjectWindow: FunctionComponent<LoadProjectWindowProps> = ({
...
+                </tbody>
+              </table>
+            </div>
+            <div>
+              <Button
+                onClick={() => importUploadedFiles({ replaceProject: true })}
+              >
+                Load into a NEW project
+              </Button>
+              <span style={{ margin: "0 10px" }}>or</span>
```
Comment (jsoules) [inline gui/src/app/pages/HomePage/LoadProjectWindow.tsx:153]: Embedded const styling

File: gui/src/app/pages/HomePage/LoadProjectWindow.tsx [hunk 3]
```diff
@@ -89,58 +81,85 @@ const LoadProjectWindow: FunctionComponent<LoadProjectWindowProps> = ({
...
-      // Otherwise, we just go ahead and import the files, replacing the
-      // entire project
-      importUploadedFiles({ replaceProject: true });
-    }
-  }, [filesUploaded, importUploadedFiles]);
+  const onUpload = useCallback(
+    (fs: { name: string; content: ArrayBuffer }[]) => {
+      if (fs.length === 1 && fs[0].name.endsWith(".zip")) {
+        setFilesUploaded(fs);
+        importUploadedFiles({ replaceProject: true });
```
Comment (magland) [inline gui/src/app/pages/HomePage/LoadProjectWindow.tsx:88]: This is not going to work. The setFilesUploaded(fs) is not going to take effect until the next render, so importUploadedFiles() should not be called immediately.

What I would suggest  is to have a separate useEffect that checks for the case of a single .zip file and then calls the importFilesUploaded()

File: gui/src/app/pages/HomePage/LoadProjectWindow.tsx [hunk 4]
```diff
@@ -49,25 +43,19 @@ const LoadProjectWindow: FunctionComponent<LoadProjectWindowProps> = ({
...
+            if (file.name.endsWith(".stan")) {
+              if (stanFileName !== "") {
+                throw Error("Only one .stan file can be uploaded at a time");
+              }
+              files["main.stan"] = parseFile(file.content);
+              stanFileName = file.name;
+              continue;
+            }
             if (!Object.values(FileNames).includes(file.name as any)) {
               throw Error(`Unrecognized file: ${file.name}`);
```
Comment (magland) [inline gui/src/app/pages/HomePage/LoadProjectWindow.tsx:60]: Suppose the user uploaded multiple files including something.zip. This message would then say "Unrecognized file: something.zip" whereas it should really say: "When importing from a Zip file, you must only upload the one .zip file" or something to that effect.

Also, I'm not sure "Unrecognized file" is the right message here. Maybe, "Unsupported file name"

**Review by magland — CHANGES_REQUESTED** [review]

File: gui/src/app/pages/HomePage/LoadProjectWindow.tsx [hunk 5]
```diff
@@ -89,58 +81,85 @@ const LoadProjectWindow: FunctionComponent<LoadProjectWindowProps> = ({
...
+          You can upload:
+          <ul>
+            <li>A .zip file that was previously exported</li>
+            <li>
+              A directory of files that were extracted from an exported .zip
+              file
+            </li>
+            <li>An individual *.stan file</li>
+            <li>
+              Other individual project files (meta.json, data.json, init.json,
```
[the remark this answers is not in the document]
Comment (WardBrian) [reply gui/src/app/pages/HomePage/LoadProjectWindow.tsx:113]: Nope, oversight by me.

File: gui/src/app/pages/HomePage/LoadProjectWindow.tsx (hunk 4 above)
Comment (WardBrian) [reply gui/src/app/pages/HomePage/LoadProjectWindow.tsx]: Updated to make both changes

File: gui/src/app/pages/HomePage/LoadProjectWindow.tsx (hunk 3 above)
Comment (WardBrian) [reply gui/src/app/pages/HomePage/LoadProjectWindow.tsx]: Yep, should have tested this! I ended up splitting up the zip logic from the function that gets called when the buttons are clicked

File: gui/src/app/pages/HomePage/LoadProjectWindow.tsx [hunk 6]
```diff
@@ -24,63 +23,64 @@ const LoadProjectWindow: FunctionComponent<LoadProjectWindowProps> = ({
...
+        }
+
+        const fileManifest = mapFileContentsToModel(files);
+        update({
+          type: "loadFiles",
+          files: fileManifest,
+          clearExisting: replaceProject,
+        });
+
+        if (stanFileName !== "" && fileManifest.meta === undefined) {
```
Comment (jsoules) [inline gui/src/app/pages/HomePage/LoadProjectWindow.tsx:80]: This is supposed to retitle the project if the .stan file uploaded is not called main.stan, right? If so I'm not sure that's working...

Comment (WardBrian) [reply gui/src/app/pages/HomePage/LoadProjectWindow.tsx:80]: It appears to be working for me? Though it should probably only do this if you hit 'new project'

Comment (jsoules) [reply gui/src/app/pages/HomePage/LoadProjectWindow.tsx]: Ah, it only functions if you don't have a meta.json in the upload. My mistake.

Comment (WardBrian) [reply gui/src/app/pages/HomePage/LoadProjectWindow.tsx]: Yeah, we let the title in `meta.json` be the definitive one in the case that there are both

**Review by magland — APPROVED** [review]

**Pull request merged**